### ISSUE TITLE
Require a setting to be provided by the environment (passwords etc.)

### ISSUE THREAD

**Issue by jonaskello** [mozilla/node-convict#251]
I'm thinking the case of a 12factor app where for example the database password is in the environment. We don't want to commit the production password to the config schema. Currently there seems to be no good way to declare that a setting is required to always be set from the enviroment although this is AFAIK best practice. In development you set it from a .env file that is not committed. This is related to #29 but seeing how that is closed I think a new issue could be useful.

**Comment by i0natan** [comment]
+1

**Comment by drzaiusx11** [comment]
+1

**Comment by madarche** [comment]
@jonaskello I agree that this would be a good feature as it would allow to follow the best practice you're mentioning.

So to anyone interested, please propose a PR for this feature, it will be very welcome!

**Comment by halfzebra** [comment]
Hey Marc!

The logical way to implement this would be to assume that the evn var is required if the `deafult` is not provided:

```js
convict({
  format: String,
  env: 'IMPORTANT_PASSWORD'
}, { env: {}}) // throws
```

But this would introduce a breaking change to the library, so the second best option is to introduce a `required` property to the schema:

```js
convict({
  format: String,
  env: 'IMPORTANT_PASSWORD',
  required: true
  default: 'wow' // this is ignored and can be omitted
}, { env: {}}) // throws
```

The `required` property would ignore the `default`.
This could also me considered as a breaking change if the `deafult` is allowed to be omitted.

What do you think?

**Comment by A-312** [comment]
Can be fixed by #313 ?

**Comment by halfzebra** [comment]
Technically yes, but for my needs it would be an overkill :muscle:

**Comment by A-312** [comment]
Why not a custom format which accept false or String ? Or default: '' ? Actually you have several solution to do that.

**Comment by jeantanzj** [comment]
```javascript
/** 
 * To require an env var
 * use
 *    default: '',
 *    format: 'required-string',
 */
 convict.addFormats({
  'required-string': {
    validate: (val: any): void => {
      if (val === '') {
        throw new Error('Required value cannot be empty')
      }
    },
    coerce: (val: any): any => {
      if (val === null) {
        return undefined
      }
      return val
    },
  })
```

If anyone else is looking for a solution
